doc/userguide/devguide/extending/app-layer/overview.rst

```diff
@@ -0,0 +1,98 @@
+**************************
+Application Layer Overview
+**************************
+
+.. contents:: Table of Contents
+
+This sections aims to give an overview of what is needed to add
+an application-layer protocol to Suricata.
+
+After a generic first step of collecting data about this application-layer protocol,
+especially pcaps for testing, we can dive into the Suricata specifics.
+
+An application-layer protocol has three logic components in Suricata :
+- parser
+- logger
+- detection engine
+
+Both detection engine and logger will depend on the processing done by the parser.
+
+Parser
+======
+
+The parser is described by an instance of the structure ``RustParser``.
+
+A parser has :
+
+- a name (where it is better to avoid dashes)
+- an ipproto (if an app-layer is both over UDP and TCP, it needs to be registered with 2 RustParser)
+- flags : only one flag ``APP_LAYER_PARSER_OPT_ACCEPT_GAPS``
+- some app-layer detection logic see :ref:`Protocol detection`.
+- some logic around one state and one transaction structures
+- some stringers functions (frames, events)
+
+So each app-layer protocol needs to define two structures : one State and one Transaction.
+A State will live throughout the flow (or until there is a protocol change in the flow).
+As such, it is useful to retain data that needs such a scope (for example HTTP2 dynamic headers table).
+And it is also useful if the parsing uses a state-machine logic, for example for file streaming.
+A State will own a list of :ref:`transactions`.
+
+Transactions
+------------
+
+Transactions are the base logical unit used by Suricata for an application-layer protocol.
+
+The big decision is how to assemble PDUs into a transaction.
+Simplest design is to have one transaction per PDU.
+But it may add value for example to combine request and response into a single transaction.
+
+The ``RustParser`` structure contains callbacks to parse network traffic in both directions.
+These callbacks will create the transactions.
+
+For protocols over TCP, we have to loop as one callback may run for a network traffic
+containing multiple PDUs, and thus resulting in the creation of multiple transactions.
+
+If a protocol may have multiple long-lived transactions, it is a good to enforce limits
```

Comment:
nit:
```suggestion
If a protocol may have multiple long-lived transactions, it is good to enforce limits
```

Reply:
Should this be made into a note, and maybe mentioned directly the `max-tx` setting or some other?